### DISCUSSION TITLE
Best way to stream data in and out?

### DISCUSSION THREAD

**Discussion by DavidDudson** [napi-rs/napi-rs#1483] [Q&A]
A simplified version of our napi rust lib would be 

Node Stream => Rust Processor => Node Stream

Essentially the Node Process needs to Write to a Stream and expose it to rust so it can be Read, processed and then Rust needs to write data to another stream so we can consume the results back from 

I'm curious about what the best practices are for dealing with this currently.

Is it possible somehow to have an napi function that looks like `process(input: Stream): Stream`

Would you 
- Pass in 2 Passthroughs and inputs, hold a reference in rust and write to them.
- Expose an napi class so that Node can read/write
- Use Threadsafe functions and some sort of event emitter style to emit buffers (avoid streams entirely)
- Use IPC tools (which we used to use, such as os pipes/sockets etc. it just feels bad considering they are the same process)

I noticed there is a crate for wasm (https://crates.io/crates/wasm-streams), but as far as I can tell there is not one for napi, is this something that could be of interest to this lib? Or am I missing something obvious.

**Comment by Brooooooklyn** [comment]
I'll implement [`Stream API`](https://developer.mozilla.org/en-US/docs/Web/API/Streams_API) after this: https://github.com/napi-rs/napi-rs/issues/796

**Reply by DavidDudson** [reply]
Awesome, glad to know it's on the roadmap. I'm happy to help out if you need a hand at all.

**Reply by justin0mcateer** [reply]
Looks like the upstream issue is resolved. Any further thoughts on Stream support?

**Reply by medemi68** [reply]
I second this. Currently need to be able to access a stream of audio data in NodeJS

**Reply by cdbattags** [reply]
Any updates here? Currently using WebSockets and could always use TCP/UDP but I wonder if a "Node.js-like" `Stream` implementation for my use-case would work better.

**Reply by remorses** [reply]
There seems to be a PR up now: https://github.com/napi-rs/napi-rs/pull/2405
